### PR TITLE
Marshal to type

### PR DESCRIPTION
**Pull request by arv**

### CODE REVIEW

**Comment by arv** [comment]
I split this into a few commits to make it clear that I'm mostly using old code (with fixes) to get the noms type from a go type.

The tests were derived from the marshal tests.

File: go/marshal/encode_type.go [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+	}
+
+	return
+}
+
+// MustMarshalType computes a Noms type from a Go type or panics if there is an
+// error.
+func MustMarshalType(v interface{}) *types.Type {
+	t, err := MarshalType(v)
+	d.Chk.NoError(err)
```
Comment (mikegray) [inline go/marshal/encode_type.go]: d.PanicIfError

File: go/marshal/encode_type.go [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+	return t
+}
+
+var typeOfTypesType = reflect.TypeOf((*types.Type)(nil))
```
Comment (mikegray) [inline go/marshal/encode_type.go]: heh

File: go/marshal/encode_type.go [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+
+type encodeTypeOptions struct {
+	IgnoreOmitempty bool
```
Comment (mikegray) [inline go/marshal/encode_type.go]: IgnoreOmitEmpty

File: go/marshal/encode_type.go [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+}
+
+func encodeType(t reflect.Type, parentStructTypes []reflect.Type, tags nomsTags, options encodeTypeOptions) *types.Type {
+	if t.Implements(marshalerInterface) {
+		// There is no way to determine the noms type now, it can be different each
+		// time MarshalNoms is called. This is handled further up the stack.
```
Comment (mikegray) [inline go/marshal/encode_type.go]: Fix comment

File: go/marshal/encode_type_test.go [hunk 5]
```diff
@@ -0,0 +1,377 @@
...
+		Type   *types.Type
+	}
+	var s S
+	assert.True(MustMarshalType(s).Equals(
+		types.MakeStructTypeFromFields("S", types.FieldMap{
+			"blob":   types.BlobType,
+			"bool":   types.BoolType,
+			"number": types.NumberType,
+			"string": types.StringType,
+			"type":   types.TypeType,
```
Comment (mikegray) [inline go/marshal/encode_type_test.go]: Include test for types.Ref?

File: go/marshal/encode_type.go (hunk 2 above)
Comment (arv) [inline go/marshal/encode_type.go]: My first name lof this was `typeTypeType` =D

File: go/marshal/encode_type_test.go (hunk 5 above)
Comment (arv) [inline go/marshal/encode_type_test.go]: None of the noms types that take type parameters are supported. I'll add tests that these errors

File: go/marshal/encode_type.go (hunk 4 above)
Comment (arv) [inline go/marshal/encode_type.go]: This is right. This is used by Marshal. I'll rephrase it slightly

File: go/marshal/encode_type.go (hunk 3 above)
Comment (arv) [inline go/marshal/encode_type.go]: Done.

File: go/marshal/encode_type.go (hunk 1 above)
Comment (arv) [inline go/marshal/encode_type.go]: Done.

**Pull request merged**